tests/returntypes/gradients/test_hadamard_grad.py

```diff
@@ -0,0 +1,853 @@
+
+        res = qml.gradients.hadamard_grad(circuit)(x)
+        assert isinstance(res, tuple)
+        assert len(res) == expected_shape[0]
+
+        if len(expected_shape) > 1:
+            for r in res:
+                assert isinstance(r, tuple)
+                assert len(r) == expected_shape[1]
+
+    costs_and_expected_probs = [
+        (cost4, [3, 4]),
+        (cost5, [3, 4]),
+        (cost6, [2, 3, 4]),
+    ]
+
+    @pytest.mark.parametrize("cost, expected_shape", costs_and_expected_probs)
+    def test_output_shape_matches_qnode_probs(self, cost, expected_shape):
+        """Test that the transform output shape matches that of the QNode."""
+        dev = qml.device("default.qubit", wires=5)
+
+        x = np.random.rand(3)
+        circuit = qml.QNode(cost, dev)
+
+        res = qml.gradients.hadamard_grad(circuit)(x)
+        assert isinstance(res, tuple)
+        assert len(res) == expected_shape[0]
+
+        if len(expected_shape) > 2:
+            for r in res:
+                assert isinstance(r, tuple)
+                assert len(r) == expected_shape[1]
+
+                for r_ in r:
+                    assert isinstance(r_, qml.numpy.ndarray)
+                    assert len(r_) == expected_shape[2]
+
+        elif len(expected_shape) > 1:
+            for r in res:
+                assert isinstance(r, qml.numpy.ndarray)
+                assert len(r) == expected_shape[1]
+
+    def test_multi_measure_no_warning(self):
+        """Test computing the gradient of a tape that contains multiple
+        measurements omits no warnings."""
+
+        dev = qml.device("default.qubit", wires=4)
+
+        par1 = qml.numpy.array(0.3)
+        par2 = qml.numpy.array(0.1)
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RY(par1, wires=0)
+            qml.RX(par2, wires=1)
+            qml.probs(wires=[1, 2])
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        with warnings.catch_warnings(record=True) as record:
+            tapes, fn = qml.gradients.hadamard_grad(tape)
+            fn(dev.batch_execute(tapes))
+
+        assert len(record) == 0
+
+
+class TestHadamardGradEdgeCases:
+    """Test the Hadamard gradient transform and edge cases such as non diff parameters, auxiliary wires, etc..."""
+
+    working_wires = [None, qml.wires.Wires(2)]
+    already_used_wires = [qml.wires.Wires(0), qml.wires.Wires(1)]
+
+    @pytest.mark.parametrize("aux_wire", working_wires)
+    def test_aux_wire(self, aux_wire):
+        """Test the aux wire is available."""
+        # One qubit is added to the device 2 + 1
+        dev = qml.device("default.qubit", wires=3)
+        x = 0.543
+        y = -0.654
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(0) @ qml.PauliX(1))
+
+        tape = qml.tape.QuantumScript.from_queue(q)
```

Comment:
I suggest we make this a QNode or a template function for the whole test class, in order to reduce code duplication.

Reply:
Did you prefer not to do so in the end?